## Importing necessary libararies for Darta Preparation 

In [32]:
import pandas as pd
import seaborn as sns
import numpy as np 
import matplotlib.pyplot as plt

## I Exploratory Data Analysis (EDA)

### 1- Loading the dataset and showing the head and tail of it.

In [33]:
df_GenWaste = pd.read_csv("Generation of waste.csv") # # Load the dataset

In [34]:
df_GenWaste.head(7) # Showing first 7 rows of data 

,Statistic Label,Year,NACE Rev. 2 Activity,Hazardousness,Waste Category,UNIT,VALUE
0,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],Total waste,Tonnes,8706.0
1,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],Primary waste (TOTAL minus SEC)[PRIM],Tonnes,NaN
2,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],Secondary waste (W033+W103+W128_13)[SEC],Tonnes,NaN
3,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],Chemical and medical wastes (subtotal)[W01-05],Tonnes,0.0
4,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],Spent solvents[W011],Tonnes,0.0
5,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],"Acid, alkaline or saline wastes[W012]",Tonnes,0.0
6,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],Used oils[W013],Tonnes,0.0


In [35]:
df_GenWaste.tail(11) # show the last 11 rows of CSV (Generation of waste)

,Statistic Label,Year,NACE Rev. 2 Activity,Hazardousness,Waste Category,UNIT,VALUE
29095,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non-hazardous[NHAZ],Combustion wastes[W124],Tonnes,128201.0
29096,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non-hazardous[NHAZ],Soils[W126],Tonnes,4276035.0
29097,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non-hazardous[NHAZ],Dredging spoils[W127],Tonnes,3.0
29098,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non-hazardous[NHAZ],Mineral wastes from waste treatment and stabil...,Tonnes,125038.0
29099,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non-hazardous[NHAZ],Metallic wastes (W061+W062+W063)[W06],Tonnes,NaN
29100,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non-hazardous[NHAZ],Animal and mixed food waste; vegetal wastes (W...,Tonnes,NaN
29101,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non-hazardous[NHAZ],Mineral waste (except non-hazardous dredging s...,Tonnes,NaN
29102,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non-hazardous[NHAZ],"Mineral wastes (except combustion wastes, cont...",Tonnes,NaN
29103,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non-hazardous[NHAZ],"Soils and dredging spoils (W126+W127, valid up...",Tonnes,NaN
29104,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non-hazardous[NHAZ],"Solidified, stabilised or vitrified wastes (va...",Tonnes,NaN


### 2- Checking some numbers of features 


In [36]:
df_GenWaste.shape # shwoing features(columns) and observations(rows) in the dataset and some other information

(29106, 7)

- This method(below code) prints a concise summary of a DataFrame, including the index range,
  
- Column names, non-null counts, data types (dtypes), and memory usage.
- When running the code, you will get the following information:
- Index Range: The total number of rows (e.g., from 0 to 29105).
- Column Count: The total number of columns(29106)
- Non-Null Count: Helps you identify if there are any missing values (NaN) in any column.
- Dtype: The data type of each column (e.g., int64 for integers, or object for text or float(64) for real numbers.
- Memory Usage: The amount of space the table occupies in the RAM(1.6+ MB)    
- We typically use "df_GenWaste.info()" at the beginning of Exploratory Data Analysis (EDA) to ensure that the data has been loaded
  correctly and to identify columns that need cleaning or data type conversion.
- Number of missing values: 11233 missing values (29106-17873=11233)
- In other words: approximately ~38.593% (11233/29106) of the data in the results column (VALUE) is missing.


In [37]:
df_GenWaste.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29106 entries, 0 to 29105
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Statistic Label       29106 non-null  object 
 1   Year                  29106 non-null  int64  
 2   NACE Rev. 2 Activity  29106 non-null  object 
 3   Hazardousness         29106 non-null  object 
 4   Waste Category        29106 non-null  object 
 5   UNIT                  29106 non-null  object 
 6   VALUE                 17873 non-null  float64
dtypes: float64(1), int64(1), object(5)
memory usage: 1.6+ MB


### 3- Statistical description

In [38]:
# The df.describe() method is used to generate descriptive statistics for DataFrame. 
# By default, it analyzes only the numeric columns (like int64 and float64) and provides a summary of their distribution
# (in this case we have just Value column)
# count: The number of non-empty (non-null) rows.

# mean: The average value.

# std: The standard deviation (how spread out the data is).

# min: The smallest value in the column.

# 25%: The 1st quartile (25% of the data falls below this value).

# 50%: The Median (the middle value of the dataset).

# 75%: The 3rd quartile (75% of the data falls below this value)

# Generate descriptive statistics to understand the distribution of numeric data.
# This helps identify the time range (2004-2020), average waste values, 
# and potential outliers or skewness in the 'VALUE' column. 
# (Understand Distribution, Spot Outliers, Data Validation)

df_GenWaste.describe() 


,Year,VALUE
count,29106.000000,1.787300e+04
mean,2012.000000,1.573629e+05
std,5.164067,1.142273e+06
min,2004.000000,0.000000e+00
25%,2008.000000,0.000000e+00
50%,2012.000000,5.400000e+01
75%,2016.000000,5.425000e+03
max,2020.000000,2.959918e+07


In [39]:
# Exploring missing values
print("Missing Values per Column:")
print(df_GenWaste.isnull().sum()) # Counting the number of missing (null) values in each column

Missing Values per Column:
Statistic Label             0
Year                        0
NACE Rev. 2 Activity        0
Hazardousness               0
Waste Category              0
UNIT                        0
VALUE                   11233
dtype: int64


In [40]:
# Calculate the total number of missing values across the entire DataFrame.
# we have 11233 missing values in the totoal , all values in the 'VALUE' Column)

df_GenWaste.isnull().sum().sum()  

np.int64(11233)

### Cleaning Data 

In [41]:

# Clean column names by removing any leading or trailing whitespaces.
df_GenWaste.columns = df_GenWaste.columns.str.strip()

# showing All new columns names( for checking previous operation)
df_GenWaste.columns

Index(['Statistic Label', 'Year', 'NACE Rev. 2 Activity', 'Hazardousness',
       'Waste Category', 'UNIT', 'VALUE'],
      dtype='object')

In [42]:
# this code Convert the 'VALUE' column to a numeric data type.
# The 'errors=coerce' parameter ensures that any non-numeric values (such as text, symbols, or empty strings)
# are transformed into NaN (Not a Number) instead of raising an error and stopping the program execution(

df_GenWaste['VALUE'] = pd.to_numeric(df_GenWaste['VALUE'], errors='coerce')
df_GenWaste.head(10)

,Statistic Label,Year,NACE Rev. 2 Activity,Hazardousness,Waste Category,UNIT,VALUE
0,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],Total waste,Tonnes,8706.0
1,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],Primary waste (TOTAL minus SEC)[PRIM],Tonnes,NaN
2,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],Secondary waste (W033+W103+W128_13)[SEC],Tonnes,NaN
3,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],Chemical and medical wastes (subtotal)[W01-05],Tonnes,0.0
4,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],Spent solvents[W011],Tonnes,0.0
5,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],"Acid, alkaline or saline wastes[W012]",Tonnes,0.0
6,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],Used oils[W013],Tonnes,0.0
7,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],Chemical wastes[W02A],Tonnes,0.0
8,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],Industrial effluent sludges[W032],Tonnes,0.0
9,Waste generated,2004,"Agriculture, forestry and fishing (A)",Hazardous and non-hazardous - Total[HAZ_NHAZ],Sludges and liquid wastes from waste treatment...,Tonnes,NaN


In [43]:
df_GenWaste['Hazardousness'].unique() # # Displaying all unique values present in the 'Hazardousness' column

array(['Hazardous and non-hazardous - Total[HAZ_NHAZ]', 'Hazardous[HAZ]',
       'Non-hazardous[NHAZ]'], dtype=object)

In [44]:


# (Clean data in the columns that will be used for merging
#( We will manually remove identifying codes (like [HAZ_NHAZ]) using a simple "replace" function)
# we replace 'Hazardous and non-hazardous - Total[HAZ_NHAZ]' by Total
# regex=False tells Pandas to treat the pattern as a literal string, not a regular expression regex (regex is language))
# an update ( We need to to remove this code below, if we want to do that for 'Hazardousness' ..... the data still non 100% cleaned)

# We replace 'Hazardous and non-hazardous - Total[HAZ_NHAZ]' by 'Non-hazardous'
df_GenWaste['Hazardousness'] = df_GenWaste['Hazardousness'].str.replace('Hazardous and non-hazardous - Total[HAZ_NHAZ]', 'Total', regex=False)

# we replace Hazardous[NHAZ]'by 'Non-hazardous'
df_GenWaste['Hazardousness'] = df_GenWaste['Hazardousness'].str.replace('Hazardous[HAZ]', 'Hazardous', regex=False)

# we replace Non-hazardous[NHAZ]'by 'Non-hazardous'
df_GenWaste['Hazardousness'] = df_GenWaste['Hazardousness'].str.replace('Non-hazardous[NHAZ]', 'Non hazardous' , regex=False)

# Cleaning 'Waste Category' column using the same simple approach) 
# Here we just trim spaces which often cause issues during merging 
# this code Cleans spaces ='whitespace' from both ends
df_GenWaste['Waste Category'] = df_GenWaste['Waste Category'].str.strip()

# Display the first few rows of the cleaned data
df_GenWaste.tail(10)

,Statistic Label,Year,NACE Rev. 2 Activity,Hazardousness,Waste Category,UNIT,VALUE
29096,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,Soils[W126],Tonnes,4276035.0
29097,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,Dredging spoils[W127],Tonnes,3.0
29098,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,Mineral wastes from waste treatment and stabil...,Tonnes,125038.0
29099,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,Metallic wastes (W061+W062+W063)[W06],Tonnes,NaN
29100,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,Animal and mixed food waste; vegetal wastes (W...,Tonnes,NaN
29101,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,Mineral waste (except non-hazardous dredging s...,Tonnes,NaN
29102,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,"Mineral wastes (except combustion wastes, cont...",Tonnes,NaN
29103,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,"Soils and dredging spoils (W126+W127, valid up...",Tonnes,NaN
29104,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,"Solidified, stabilised or vitrified wastes (va...",Tonnes,NaN
29105,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,Waste excluding major mineral wastes[TOT_X_MIN],Tonnes,6650806.0


In [45]:

df_GenWaste.columns # Display the names of all columns in the DataFrame

Index(['Statistic Label', 'Year', 'NACE Rev. 2 Activity', 'Hazardousness',
       'Waste Category', 'UNIT', 'VALUE'],
      dtype='object')

In [46]:
# Drop the 'UNIT' column and rename 'VALUE' column to 'VALUE(Tonnes)' for simplicity.
df_GenWaste = df_GenWaste.drop(columns=['UNIT']).rename(columns={'VALUE': 'VALUE(Tonnes)'})


In [47]:
df_GenWaste

,Statistic Label,Year,NACE Rev. 2 Activity,Hazardousness,Waste Category,VALUE(Tonnes)
0,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Total waste,8706.0
1,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Primary waste (TOTAL minus SEC)[PRIM],NaN
2,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Secondary waste (W033+W103+W128_13)[SEC],NaN
3,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Chemical and medical wastes (subtotal)[W01-05],0.0
4,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Spent solvents[W011],0.0
...,...,...,...,...,...,...
29101,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,Mineral waste (except non-hazardous dredging s...,NaN
29102,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,"Mineral wastes (except combustion wastes, cont...",NaN
29103,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,"Soils and dredging spoils (W126+W127, valid up...",NaN
29104,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,"Solidified, stabilised or vitrified wastes (va...",NaN


In [48]:
# # Printing all unique values in the 'Waste Category' column
print(df_GenWaste['Waste Category'].unique())

['Total waste' 'Primary waste (TOTAL minus SEC)[PRIM]'
 'Secondary waste (W033+W103+W128_13)[SEC]'
 'Chemical and medical wastes (subtotal)[W01-05]' 'Spent solvents[W011]'
 'Acid, alkaline or saline wastes[W012]' 'Used oils[W013]'
 'Chemical wastes[W02A]' 'Industrial effluent sludges[W032]'
 'Sludges and liquid wastes from waste treatment[W033]'
 'Health care and biological wastes[W05]'
 'Recyclable wastes (subtotal, W06+W07 except W077)[W06_07A]'
 'Metal wastes, ferrous[W061]' 'Metal wastes, non-ferrous[W062]'
 'Metal wastes, mixed ferrous and non-ferrous[W063]' 'Glass wastes[W071]'
 'Paper and cardboard wastes[W072]' 'Rubber wastes[W073]'
 'Plastic wastes[W074]' 'Wood wastes[W075]' 'Textile wastes[W076]'
 'Equipment (subtotal, W077+W08A+W081+W0841)[W077_08]'
 'Waste containing PCB[W077]'
 'Discarded equipment (except discarded vehicles and batteries and accumulators waste) (W08 except W081, W0841)[W08A]'
 'Discarded vehicles[W081]' 'Batteries and accumulators wastes[W0841]'
 'Animal 

In [49]:
# # Printing  all unique values in the 'NACE Rev. 2 Activity' column
print(df_GenWaste['NACE Rev. 2 Activity'].unique())


['Agriculture, forestry and fishing (A)' 'Mining and quarrying (B)'
 'Manufacturing (C)'
 'Manufacture of food products; beverages and tobacco products (C10-C12)'
 'Manufacture of textiles, wearing apparel, leather and related products  (C13-C15)'
 'Manufacture of wood and of products of wood and cork, except furniture; manufacture of articles of straw and plaiting materials (C16)'
 'Manufacture of paper and paper products; printing and reproduction of recorded media (C17_C18)'
 'Manufacture of coke and refined petroleum products (C19)'
 'Manufacture of chemical, pharmaceutical, rubber and plastic products (C20-C22)'
 'Manufacture of other non-metallic mineral products (C23)'
 'Manufacture of basic metals and fabricated metal products, except machinery and equipment  (C24_C25)'
 'Manufacture of computer, electronic and optical products, electrical equipment, motor vehicles and other transport equipment (C26-C30)'
 'Manufacture of furniture; jewellery, musical instruments, toys; repair 

In [50]:
# Final Trim to remove any hidden spaces after replacement
# After replacing, we strip again to make sure no spaces were left behind
df_GenWaste['Waste Category'] = df_GenWaste['Waste Category'].str.strip()

# Convert VALUE to numeric (Coerce errors will handle missing data)
df_GenWaste['VALUE(Tonnes)'] = pd.to_numeric(df_GenWaste['VALUE(Tonnes)'], errors='coerce')


In [51]:
df_GenWaste

,Statistic Label,Year,NACE Rev. 2 Activity,Hazardousness,Waste Category,VALUE(Tonnes)
0,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Total waste,8706.0
1,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Primary waste (TOTAL minus SEC)[PRIM],NaN
2,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Secondary waste (W033+W103+W128_13)[SEC],NaN
3,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Chemical and medical wastes (subtotal)[W01-05],0.0
4,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Spent solvents[W011],0.0
...,...,...,...,...,...,...
29101,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,Mineral waste (except non-hazardous dredging s...,NaN
29102,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,"Mineral wastes (except combustion wastes, cont...",NaN
29103,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,"Soils and dredging spoils (W126+W127, valid up...",NaN
29104,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,"Solidified, stabilised or vitrified wastes (va...",NaN


In [52]:
print(df_GenWaste.dtypes) # we don't need to make VALUE columns as float (already float) that its= shows in the VALUE float64 but it still problem
                          # the next code will show us the problem 

Statistic Label          object
Year                      int64
NACE Rev. 2 Activity     object
Hazardousness            object
Waste Category           object
VALUE(Tonnes)           float64
dtype: object


In [53]:
# Convert VALUE column to numeric
# errors='coerce' will turn any non-numeric value (like empty strings) into NaN
# errors='coerce': This is the "safety net." If Pandas encounters something it can't turn into a number (like the word "Unknown" or a typo),
# it forces it to become NaN.
# It cleans our data and prepares the column for mathematical operations (like .sum() or .mean(), that's Why we use it)

df_GenWaste['VALUE(Tonnes)'] = pd.to_numeric(df_GenWaste['VALUE(Tonnes)'], errors='coerce')


In [54]:
# save the brut data in Cleaned Generation of Waste NaN=0
df_GenWaste.to_csv("Cleaned Generation of Waste_0.csv")


In [55]:
#  We replace missing values in 'VALUE(Tonnes)' column with 0
df_GenWaste['VALUE(Tonnes)'] = df_GenWaste['VALUE(Tonnes)'].fillna(0)

# we save the DataFrame to a new CSV file called "Cleaned Generation of Waste NaN=0" without the index column.
df_GenWaste.to_csv("Cleaned Generation of Waste NaN=0.csv")

# We Save the file in a new name (with 0 in empty boxes of Values(Tonnes)
df_Cleaned_GenWaste_0= pd.read_csv("Cleaned Generation of Waste NaN=0.csv")

In [56]:
df_GenWaste

,Statistic Label,Year,NACE Rev. 2 Activity,Hazardousness,Waste Category,VALUE(Tonnes)
0,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Total waste,8706.0
1,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Primary waste (TOTAL minus SEC)[PRIM],0.0
2,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Secondary waste (W033+W103+W128_13)[SEC],0.0
3,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Chemical and medical wastes (subtotal)[W01-05],0.0
4,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Spent solvents[W011],0.0
...,...,...,...,...,...,...
29101,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,Mineral waste (except non-hazardous dredging s...,0.0
29102,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,"Mineral wastes (except combustion wastes, cont...",0.0
29103,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,"Soils and dredging spoils (W126+W127, valid up...",0.0
29104,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,"Solidified, stabilised or vitrified wastes (va...",0.0


In [57]:
df_Cleaned_GenWaste_0

,Unnamed: 0,Statistic Label,Year,NACE Rev. 2 Activity,Hazardousness,Waste Category,VALUE(Tonnes)
0,0,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Total waste,8706.0
1,1,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Primary waste (TOTAL minus SEC)[PRIM],0.0
2,2,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Secondary waste (W033+W103+W128_13)[SEC],0.0
3,3,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Chemical and medical wastes (subtotal)[W01-05],0.0
4,4,Waste generated,2004,"Agriculture, forestry and fishing (A)",Total,Spent solvents[W011],0.0
...,...,...,...,...,...,...,...
29101,29101,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,Mineral waste (except non-hazardous dredging s...,0.0
29102,29102,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,"Mineral wastes (except combustion wastes, cont...",0.0
29103,29103,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,"Soils and dredging spoils (W126+W127, valid up...",0.0
29104,29104,Waste generated,2020,All NACE activities plus households (TOTAL_HH),Non hazardous,"Solidified, stabilised or vitrified wastes (va...",0.0


### Remark
- I can clean also the names by removing between [] using regex (dynamic patern) but i don't need that for Statistics and Machine learning    project but if I needed it if I wanted to merge two files and work with the data, or if I wanted to work with secondary waste and primary   waste( that’s depends de project that that I want to do)
   However, some features(rows) are not classified consistently between the years 2004 and 2010 approximately, and some datas are missing.
In that case, I could fill the missing values using the mean, and it would be better to fill them using the average between the previous and the next values (interpolation). However, I changed the project.”


### Reference 
- www.google.com 
- Tutorials, PDF and IPTNB file of CCT - Professor: David
- Python for Data Analysis - Third edition - Wes McKinney
- https://moodle.cct.ie/course
- https://data.cso.ie/table/GWA01
- Gemeni.ai (to deep understand of regex as langage, and "regex" word)